### PR TITLE
Support methods from external classes in @MethodSource #1088

### PR DESCRIPTION
**Pull request by karollewandowski**
I cleaned by the way unused `throws` clauses in `ReflectionUtilsTests`.
FQMN parsing result is just a String array, I think it's simple enough and creating something like `MethodInfo` data class would be too much.
Please carefully review added test cases - I didn't have too much time for it and could miss some important ones.

### CODE REVIEW

**Comment by sbrannen** [comment]
Please add `Issue: #1088` to the bottom of each commit message.

**Comment by sbrannen** [comment]
> I cleaned by the way unused throws clauses in ReflectionUtilsTests.

Would you mind doing one of the following?

- submitting a separate PR for that
- moving those changes to a separate commit within the existing PR
- allowing one of the team to make those changes on `master`

**Comment by karollewandowski** [comment]
@sbrannen, I squashed everything to single commit and reverted cleaning unnecessary `throws` clauses. I noticed there is no convention to have it everywhere and just wanted to be a good scout ;-) I'm leaving it for the team.

**Comment by karollewandowski** [comment]
I just realized that I forgot to include new changes in docs. Should I do it or docs are updated in scope of separate issues?

**Comment by sbrannen** [comment]
> @sbrannen, I squashed everything to single commit and reverted cleaning unnecessary throws clauses. I noticed there is no convention to have it everywhere and just wanted to be a good scout ;-) I'm leaving it for the team.

Sounds good

**Comment by sbrannen** [comment]
> I just realized that I forgot to include new changes in docs. Should I do it or docs are updated in scope of separate issues?

Updates to the User Guide, Release Notes, and JavaDoc should be part of the PR.

**Comment by karollewandowski** [comment]
@sbrannen, docs updated.
There is a problem with Spotlight analysis. I created `documentation/src/test/java/example/ExternalMethodSourceTestDemo.java` to demonstrate factory method from external class, but Spotlight gets violation on tag declaration:
```
...
// tag::external_MethodSource_example[]    <-- Spotlight removes it
+package example;
...
```
I want to put it before package to clearly demonstrate full FQNM with package name. I tried to put `//@formatter:off` before tag, but didn't help. Any ideas?

**Comment by sbrannen** [comment]
> @sbrannen, docs updated.

Thanks!

We'll have a look...


> I want to put it before package to clearly demonstrate full FQNM with package name. I tried to put //@formatter:off before tag, but didn't help. Any ideas?

Yeah: it simply won't work. 😉 

Spotless requires that the copyright header be the first and only thing before the `package` declaration line in any given source file.

So the only option is to mention the package explicitly in the prose describing the example.

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.2.0-M1.adoc [hunk 1]
```diff
@@ -39,7 +39,8 @@ on GitHub.
 
 ==== New Features and Improvements
 
-* ❓
+* `@MethodSource` now supports external classes factory methods referenced by fully
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.2.0-M1.adoc]: Please change to:

`@MethodSource` now supports `static` factory methods declared in external classes referenced by _fully qualified method name_.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 2]
```diff
@@ -841,10 +841,11 @@ include::{testDir}/example/ParameterizedTestDemo.java[tags=EnumSource_regex_exam
 [[writing-tests-parameterized-tests-sources-MethodSource]]
 ===== @MethodSource
 
-`@MethodSource` allows you to refer to one or more _factory_ methods of the test class.
-Such methods must return a `Stream`, `Iterable`, `Iterator`, or array of arguments. In
-addition, such methods must not accept any arguments. By default such methods must be
-`static` unless the test class is annotated with `@TestInstance(Lifecycle.PER_CLASS)`.
+`@MethodSource` allows you to refer to one or more _factory_ methods of the test or
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: "of the test or external class" --> "of the test class or external classes"

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 3]
```diff
@@ -841,10 +841,11 @@ include::{testDir}/example/ParameterizedTestDemo.java[tags=EnumSource_regex_exam
...
+external class. Such methods must return a `Stream`, `Iterable`, `Iterator`, or array
+of arguments. In addition, such methods must not accept any arguments. By default such
+methods must be `static` unless the test class is annotated with
+`@TestInstance(Lifecycle.PER_CLASS)`.
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: The last sentence is no longer technically correct.

Factory methods in external classes must always be `static`.

Thus, please reword that to take into account the change in this PR.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 4]
```diff
@@ -880,6 +881,14 @@ factory method defined in the `Arguments` interface.
 include::{testDir}/example/ParameterizedTestDemo.java[tags=multi_arg_MethodSource_example]
 ----
 
+External _factory_ method can be referenced by providing _fully qualified method name_ as
+demonstrated in the following example.
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc]: Please change to:

An external, `static` _factory_ method can be referenced by providing its _fully qualified method name_ as
+demonstrated in the following example.

File: documentation/src/test/java/example/ExternalMethodSourceTestDemo.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+// tag::external_MethodSource_example[]
+package example;
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ExternalMethodSourceTests {
```
Comment (sbrannen) [inline documentation/src/test/java/example/ExternalMethodSourceTestDemo.java]: Please rename both the _file_ and the _class_ to: `ExternalMethodSourceDemo`.

Example classes in the `documentation` project must be named "*Demo".

File: documentation/src/test/java/example/ExternalMethodSourceTestDemo.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+// tag::external_MethodSource_example[]
```
Comment (sbrannen) [inline documentation/src/test/java/example/ExternalMethodSourceTestDemo.java]: Please move the _tag_ declaration below the `package` declaration. See comments in PR.

**Comment by sbrannen** [comment]
@karollewandowski, are you comfortable with the current _implementation_ in this PR?

In other words, aside from the documentation, do you feel this PR is ready for a final review?

**Comment by karollewandowski** [comment]
> So the only option is to mention the package explicitly in the prose describing the example within the User Guide.

I inlined `package example;` just before file include in `writing-tests.adoc`. If it's too hacky, please let me know to revert it.

> I requested a few minor changes for the documentation.

All comments resolved, please take a look again.

> @karollewandowski, are you comfortable with the current implementation in this PR?

Yes, please review it.

**Comment by sbrannen** [comment]
@karollewandowski, thanks for making changes.

I'll make sure to give it a final review and/or merge it in the next day or so.

**Comment by sbrannen** [comment]
@karollewandowski, 

FYI: I'd hoped to get to _this_ issue today, but I just lost quite a bit of time with this super annoying 🐛: #1345

**Comment by sbrannen** [comment]
This has been merged into `master` in f415edc742cda271f3c1f5c5e9b20247e92db87b and minorly polished in 13de8c51701b69769328b55b1857631e75bc5cdc.

Cheers!

**Pull request closed without merging**